### PR TITLE
fixes to enable all LTP waitpid tests

### PR DESCRIPTION
**Pull request by paulcallen**
* Implement SIGSTOP/SIGCONT to stop and resume all threads in a process
* Implement waitpid WUNTRACED flag (waitpid returns if have stopped processes)
* Fixed canary on forked processes,
* Fix issue when process shuts down due to exception, it now  makes sure it waits for child forked processes if present

Signed-off-by: Paul Allen <paul.c.allen@microsoft.com>

### CODE REVIEW

File: crt/exit.c [hunk 1]
```diff
@@ -26,27 +26,37 @@ void exit(int code)
...
+             * do crt cleanup. Children cannot because they are sharing the CRT
+             * with the parent and the parent already has a lot of state set up
+             * for them. If the child shuts down the crt the parent will crash
+             * as well as any siblings.
+             */
+
+            /* Because we are not going to do CRT cleanup we do need to make
+             * sure files are flushed before they completely shutdown. This
+             * would normally happen in the CRT cleanup code. */
+            fflush(NULL);
```
Comment (mikbras) [inline crt/exit.c:45]: Cool! I did not know about passing null to fflush.

File: crt/fork.c [hunk 2]
```diff
@@ -122,13 +122,13 @@ struct pthread* _create_child_pthread_and_copy_stack(
...
-              MAP_ANONYMOUS,
-              -1,
-              0)))
+    if ((map = mmap(
+             NULL,
+             size_rounded,
+             PROT_READ | PROT_WRITE,
+             MAP_ANONYMOUS,
+             -1,
+             0)) == MAP_FAILED)
```
Comment (mikbras) [inline crt/fork.c:131]: Ooh! Good catch. I missed that before.

File: crt/fork.c [hunk 3]
```diff
@@ -122,13 +122,13 @@ struct pthread* _create_child_pthread_and_copy_stack(
...
-              size_rounded,
-              PROT_READ | PROT_WRITE,
-              MAP_ANONYMOUS,
-              -1,
-              0)))
+    if ((map = mmap(
+             NULL,
+             size_rounded,
+             PROT_READ | PROT_WRITE,
+             MAP_ANONYMOUS,
```
Comment (mikbras) [inline crt/fork.c:129]: Suggest adding ``MAP_PRIVATE`` too.

File: crt/fork.c [hunk 4]
```diff
@@ -299,17 +300,26 @@ static int _fixup_frame_pointers(
...
+    return __pthread_self()->canary;
+}
+
 static int _child_func(void* arg)
 {
     struct thread_args* args = (struct thread_args*)arg;
     args->env.rsp = (uint64_t)args->child_sp;
     args->env.rbp = (uint64_t)args->child_bp;
 
+    _set_fsbase(args->child_pthread);
```
Comment (mikbras) [inline crt/fork.c:314]: Why does this have to go first?

File: kernel/mman.c [hunk 5]
```diff
@@ -638,11 +638,14 @@ static int _munmap(myst_mman_t* mman, void* addr, size_t length)
     // perform scrub efficiently. The prot_vector based implementation has
     // complication of inconsistent prot within a VAD. Skip scrubbing for
     // now.
-#if 0
```
Comment (mikbras) [inline kernel/mman.c:641]: Maybe you removed this by mistake?

File: tests/mman/test_mman.c [hunk 6]
```diff
@@ -20,6 +21,8 @@
 #define PAGE_SIZE 4096
 #endif
 
+myst_kernel_args_t __myst_kernel_args;
```
Comment (mikbras) [inline tests/mman/test_mman.c:24]: Is this used anywhere?

**Review by mikbras — COMMENTED** [review]
Just a few comments. Still getting started.

File: kernel/mman.c (hunk 5 above)
Comment (paulcallen) [reply kernel/mman.c:641]: It was broken before and I kind of figured the current solution was better. I can add both checks though.

File: tests/mman/test_mman.c (hunk 6 above)
Comment (paulcallen) [reply tests/mman/test_mman.c:24]: lol, yes. The test does not build without it as the kernel needs it to be there

File: crt/fork.c (hunk 3 above)
Comment (paulcallen) [reply crt/fork.c:129]: ok

Comment (vtikoo) [reply crt/fork.c:129]: @paulcallen you've added MAP_PRIVATE to the `prot`, instead of `flags`!

**Review by mikbras — APPROVED** [review]

**Pull request merged**